File: python/logisticRegression/logisticRegression.py
from numpy import *

#读取数据
def loadDataSet(filename):
    dataMat = []; labelMat = []
    fr = open(filename)
    for line in fr.readlines():
        lineArr = line.strip().split()
        dataMat.append([1.0, float(lineArr[0]), float(lineArr[1])]) #将x0的值设置为1.0
        labelMat.append(int(lineArr[2]))
    return dataMat,labelMat

#Sigmoid函数
def sigmoid(inX):
    return 1.0/(1+exp(-inX))


def gradAscent(dataMatIn, classLabels):
    dataMatrix = mat(dataMatIn)             #convert to NumPy matrix
    labelMat = mat(classLabels).transpose() #convert to NumPy matrix
    m,n = shape(dataMatrix)
    alpha = 0.001
    maxCycles = 500
    weights = ones((n,1))
    for k in range(maxCycles):              #heavy on matrix operations
        h = sigmoid(dataMatrix*weights)     #matrix mult
        error = (labelMat - h)              #vector subtraction
        weights = weights + alpha * dataMatrix.transpose()* error #matrix mult
    return weights


def plotBestFit(weights,dataMat,labelMat):
    import matplotlib.pyplot as plt
    dataArr = array(dataMat)
    n = shape(dataArr)[0]
    xcord1 = []; ycord1 = []
    xcord2 = []; ycord2 = []
    for i in range(n):
        if int(labelMat[i])== 1:
            xcord1.append(dataArr[i,1]); ycord1.append(dataArr[i,2])
        else:
            xcord2.append(dataArr[i,1]); ycord2.append(dataArr[i,2])

    fig = plt.figure()
    ax = fig.add_subplot(111)
    ax.scatter(xcord1, ycord1, s=30, c='red', marker='s')
    ax.scatter(xcord2, ycord2, s=30, c='green')
    x = arange(-3.0, 3.0, 0.1)
    print(x)
    y = (-weights[0]-weights[1]*x)/weights[2]
    print(y)
    ax.plot(x, y)
    plt.xlabel('X1'); plt.ylabel('X2');
    plt.show()

File: matlab/logisticRegression/logisticRegression.m
function [weights,dataSet,labelMat]=logisticRegression(filename)
    format long
    [dataSet,labelMat]=loadDataSet(filename);
    [m,n]=size(dataSet);
    %alpha = 0.001;  %向目标移动的步长
    maxCycles = 500;    %迭代次数
    weights = ones(1,n);
    for j=1:maxCycles
        dataIndex=1:1:m;
        for i=1:m
            if(isempty(dataIndex))
                break;
            end
            alpha = 4/(1.0+j+i)+0.0001;
            randIndex=ceil(rand(1,1)*length(dataIndex));
            h=sigmoid(weights*dataSet(randIndex,:).');
            error=(labelMat(randIndex,:)-h);
            weights = weights + alpha * error * dataSet(randIndex,:);
            dataIndex(:,randIndex)=[];
        end
    end

end

%读取文本数据，以矩阵形式返回
function [dataSet,labelMat]=loadDataSet(filename)
    format long
    dataMatTemp=importdata(filename,'\t');  %将文件内容读取到一个矩阵
    [m,n]=size(dataMatTemp);
    dataSet=zeros(m,n);
    dataSet(:,1)=1;
    dataSet(:,2:n)=dataMatTemp(:,1:n-1);
    labelMat=dataMatTemp(:,n);
end

%Sigmoid函数
function sigmoidre=sigmoid(x)
    format long
    sigmoidre=1.0./(1+exp(-x));
end



